### PR TITLE
Improve Feature Registry Usability

### PR DESCRIPTION
**Pull request by xiaoyongzhu**
Goal of this PR is to solve some critical issues that the previous PR by @YihuiGuo didn't address. More specifically, it includes:
- Getting all the components from the registry back, including the correct keys, Transformation functions, UDFs, etc.
- Adding a `feature_config` field in the schema to enable further extensibility 
- Fix a few grammar issues in Python objects, otherwise they won't pass pylint
- Adding test to make sure that the features returned from the registry can actually run a job

### CODE REVIEW

File: feathr_project/feathr/_feature_registry.py [hunk 1]
```diff
@@ -915,18 +920,80 @@ def search_input_anchor_features(self,derived_guids,feature_entity_guid_mapping)
         return result
 
 
-
+    def _correct_function_identation(self, user_func: str) -> str:
+        """The function read from registry might have the wrong identation. We need to correct those identations.
```
Comment (hangfei) [inline feathr_project/feathr/_feature_registry.py:931]: could you explain why we got wrong indentation in the doc as well.

File: feathr_project/feathr/_feature_registry.py [hunk 2]
```diff
@@ -915,18 +920,80 @@ def search_input_anchor_features(self,derived_guids,feature_entity_guid_mapping)
         return result
 
 
-
+    def _correct_function_identation(self, user_func: str) -> str:
```
Comment (hangfei) [inline feathr_project/feathr/_feature_registry.py:930]: put this in the preprocessing_py_udf module?

File: feathr_project/feathr/_feature_registry.py [hunk 3]
```diff
@@ -915,18 +920,80 @@ def search_input_anchor_features(self,derived_guids,feature_entity_guid_mapping)
...
+        # strip the lines to make sure the function has the correct indentation
+        udf_source_code_striped = [line[leading_space_num:] for line in temp_udf_source_code]
+        # append '\n' back since it was deleted due to the previous split
+        udf_source_code = [line+'\n' for line in udf_source_code_striped]
+        return " ".join(udf_source_code)
 
     def _get_source_by_guid(self, guid) -> Source:
         # TODO: currently return HDFS source by default. For JDBC source, it's currently implemented using HDFS Source so we should split in the future
         source_entity = self.purview_client.get_entity(guid=guid)["entities"][0]
+        print(source_entity["attributes"]["preprocessing"],)
```
Comment (hangfei) [inline feathr_project/feathr/_feature_registry.py:948]: remove print?

File: feathr_project/feathr/client.py [hunk 4]
```diff
@@ -700,8 +700,10 @@ def _get_kafka_config_str(self):
             """.format(sasl=sasl)
         return config_str
 
-    def get_features_from_registry(self,project_name):
+    def get_features_from_registry(self, project_name: str):
```
Comment (hangfei) [inline feathr_project/feathr/client.py:703]: change get to import or load?

File: feathr_project/feathr/_preprocessing_pyudf_manager.py [hunk 5]
```diff
@@ -58,9 +66,41 @@ def build_anchor_preprocessing_metadata(anchor_list, local_workspace_dir):
...
+            # find the first object, that is the str, this will be the name of the function
+            if isinstance(ele, str):
+                return ele
+
+
     @staticmethod
     def persist_pyspark_udf_to_file(user_func, local_workspace_dir):
-        udf_source_code = inspect.getsourcelines(user_func)[0]
+        """persist the pyspark UDF to a file in `local_workspace_dir` for later usage. 
+        The user_func could be either a string that represents a function body, or a callable object.
```
Comment (hangfei) [inline feathr_project/feathr/_preprocessing_pyudf_manager.py:91]: explain why it can be a string.

File: feathr_project/feathr/client.py (hunk 4 above)
Comment (jaymo001) [inline feathr_project/feathr/client.py:703]: Change the return type to a Dict[str, FeatureBase] ? see https://github.com/linkedin/feathr/pull/232/files#diff-6bac58e5a2ddd48b771dc23fdc033617dfa12b1ef25bbfe825806a161c543e52

File: feathr_project/feathr/_feature_registry.py (hunk 2 above)
Comment (xiaoyongzhu) [reply feathr_project/feathr/_feature_registry.py:934]: I thought about it as well but think it's cleaner to put it in the registry module, i.e. the functions that are retrieved from registry should have the correct indentation.

File: feathr_project/feathr/client.py (hunk 4 above)
Comment (xiaoyongzhu) [reply feathr_project/feathr/client.py]: let's talk offline on this

**Review by hangfei — DISMISSED** [review]

**Review by hangfei — APPROVED** [review]

**Review by jaymo001 — APPROVED** [review]

**Pull request merged**